[AERIE-2119] Structure `failWith()` reasons

pcrosemurgy opened the pull request:
* **Tickets addressed:** AERIE-2119
* **Review:** By commit  <!-- Choose from: "by commit", "by file" -->
* **Merge strategy:** Merge (no squash)  <!-- Choose from: "merge (no squash)", "squash and merge" -->

## Description

Motivated by [discussion](https://jpl.slack.com/archives/GJ07FGRPH/p1664308622371219) with @camargo and @duranb. The UI and API clients would like to have some structure within our `simulation_dataset.reason` and `scheduling_request.failure_reason` columns.

This PR introduces a serialization of these `reason`s with `ScheduleFailure` and `SimulationFailure`. Notably, there's now a `message` field to act as a human-readable summary of the failure.

This PR also renames `scheduling_request.failure_reason` -> `scheduling_request.reason`.

## Verification

- End-to-end tests passing
- Manual inspecting `simulation_dataset.reason` and `scheduler_request.failure_reason` columns after end-to-end tests.

After simulating with an activity (`BiteBanana`) with extraneous arguments `simulation_dataset.reason` contains:
```json
{
    "type": "PLAN_CONTAINS_UNCONSTRUCTABLE_ACTIVITIES",
    "message": "Plan contains unconstructable activities",
    "data": {
        "success": false,
        "errors": {
            "2": {
                "reason": {
                    "success": false,
                    "errors": {
                        "extraneousArguments": [
                            "bad_three",
                            "bad_one",
                            "bad_two"
                        ],
                        "unconstructableArguments": [],
                        "missingArguments": []
                    },
                    "arguments": {
                        "biteSize": 1.0
                    }
                }
            }
        }
    },
    "trace": "",
    "timestamp": "2022-10-10T18:05:27.5648916Z.564891"
}

```

@camargo / @duranb: Note that the inner `data.errors` in this case is now the exact same structure that's reported from `getActivityEffectiveArguments`/`getModelEffectiveArguments`. This should hopefully make ingesting this data easier.

## Documentation

TODO: will update this section once this PR has general/informal approval.

## Future work
- Find other services/areas where our error reporting could be more human-friendly.

pcrosemurgy committed:
Use `ScheduleFailure` for scheduler failure reasons

Broadly, this also required changing:
- `scheduling_request.failure_reason` to a `jsonb` type
- The `failWith()` interface and any associated callers of this method

pcrosemurgy commented on scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/SynchronousSchedulerAgent.java:
```diff
@@ -169,7 +173,10 @@ public void schedule(final ScheduleRequest request, final ResultsProtocol.Writer
       if (!failedGoals.isEmpty()) {
-        writer.failWith(failedGoals.toString());
+        writer.failWith(b -> b
+            .type("SCHEDULING_GOALS_FAILED")
```
It's maybe not ideal that these `type` fields are hardcoded string literals but I figured this is pretty innocuous right now as
1. there are no duplicate failure types right now, i.e., there's only ever one `failWith()` call per failure type.
1. every `failWith()` invocation is within this method – with the exception of one `failWith()` within the `ThreadedSchedulerAgent` that is called as a catch-all for any `Throwable`.

mattdailis replied:
> there's only ever one failWith() call per failure type

I'm in favor of this property because I think it will help make the errors more traceable 👍

pcrosemurgy committed:
Update e2e-tests `SchedulingResponse.reason` type

pcrosemurgy committed:
Use `SimulationFailure` for sim failure reasons

Broadly, this also required changing:
- `simulation_dataset.reason` to a `jsonb` type
- The `failWith()` interface and any associated callers of this method

mattdailis commented on scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/ThreadedSchedulerAgent.java:
```diff
@@ -64,7 +64,9 @@ public void run() {
               ex.printStackTrace(System.err);
-              req.writer().failWith(ex);
+              req.writer().failWith(b -> b
+                  .type("UNEXPECTED_SCHEDULER_EXCEPTION")
```
🔵 I'm wondering if it's worth having a message here - "Something went wrong while scheduling" - I could go either way.

pcrosemurgy replied:
That's a good idea, just pushed a change to add a `.message()` here.

mattdailis reviewed (approved):
Looks good to me overall

mattdailis commented on scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/ScheduleFailure.java:
```diff
@@ -0,0 +1,45 @@
+      return this;
+    }
+
+    public Builder data(final JsonValue context) {
```
🔵 Curious about the choice to name the method `data` but the field `context` - any reason for them to be different?

pcrosemurgy replied:
Oops, good catch! Just changed this to `data`.

pcrosemurgy committed:
Report strucutured activity instantiation failures

Previously "no such activity type" and "invalid arguments" activity
instantiation errors were just converted to a string and returned
through the Merlin server endpoint.

With the help of a sealed type we're now able to return the correct structure
within our response serializers depending on what the underlying error is.

duranb reviewed (commented):
@pcrosemurgy This looks great! One thing that we forgot to realize is that the design has timestamps for the errors. Is that something that is easy to add to this?

Also, I haven't done any Java in yeeears so I might sound crazy, but does it make sense to make the `Failure` records implement some base generic `Failure` interface for when we eventually add more `Failure` types?

pcrosemurgy commented:
@duranb

>One thing that we forgot to realize is that the design has timestamps for the errors. Is that something that is easy to add to this?

Ahh okay! I'll add a timestamp field to the `simulation_dataset.reason` and `scheduling_request.failure_reason` columns.

>does it make sense to make the Failure records implement some base generic Failure interface for when we eventually add more Failure types?

So with this PR `SimulationFailure` and `ScheduleFailure` serve as this interface because every `writer.failWith()` call (anything that can write to the `reason` column) must build a `SimulationFailure`/`ScheduleFailure` (depending on which server we're talking about). :)

pcrosemurgy committed:
Add timestamp to `SimulationFailure` & `ScheduleFailure`

duranb commented:
@pcrosemurgy 
> Ahh okay! I'll add a timestamp field to the `simulation_dataset.reason` and `scheduling_request.failure_reason` columns.

Thank you!

> So with this PR `SimulationFailure` and `ScheduleFailure` serve as this interface because every `writer.failWith()` call (anything that can write to the `reason` column) must build a `SimulationFailure`/`ScheduleFailure` (depending on which server we're talking about). :)

Got it! Thanks for clearing that up for me!

pcrosemurgy commented:
Verification section has been updated to show this new `timestamp` as a field within the reason JSON

pcrosemurgy committed:
Remove  assignment in claim

duranb reviewed (approved):
These changes are good with me 💯, but I'd like to see if @camargo has any thoughts.

pcrosemurgy committed:
Rename `scheduling_request.failure_reason` -> `scheduling_request.reason`
